respy/state_space.py

```diff
@@ -1,358 +1,245 @@
+        optim_paras,
+        options,
+    )
 
-    """
+    return state_space
 
-    def _create_slices_by_core_periods(self):
-        """Create slices to index all attributes in a given period.
 
-        It is important that the returned objects are not fancy indices. Fancy indexing
-        results in copies of array which decrease performance and raise memory usage.
+class StateSpaceClass:
+    """Explain how things work once finally decided upon."""
 
+    def __init__(
+        self,
+        core,
+        indexer,
+        dense,
+        dense_period_cores,
+        core_index_to_complex,
+        core_index_to_indices,
+        optim_paras,
+        options,
+    ):
         """
-        period = self.core.period
-        indices = np.where(period - period.shift(1).fillna(-1) == 1)[0]
-        indices = np.append(indices, self.core.shape[0])
-
-        slices = [slice(indices[i], indices[i + 1]) for i in range(len(indices) - 1)]
-
-        return slices
-
-    def _create_is_inadmissible(self, optim_paras, options):
-        core = self.core.copy()
-
-        for choice in optim_paras["choices"]:
-            core[choice] = False
-            for formula in options["inadmissible_states"][choice]:
-                core[choice] |= core.eval(formula)
-
-        is_inadmissible = core[optim_paras["choices"]].to_numpy(dtype=np.bool)
+        """
+        self.core = core
+        self.indexer = indexer
+        self.dense_period_cores = dense_period_cores
+        self.dense = dense
+        self.core_index_to_complex = core_index_to_complex
+        self.core_index_to_indices = core_index_to_indices
+        self.optim_paras = optim_paras
+        self.options = options
+        self.set_convenience_objects()
+        self.child_indices = self.collect_child_indices()
+        self.base_draws_sol = self.create_draws()
+        self.expected_value_functions = {
+            key: np.zeros(len(self.index_to_indices[key]))
+            for key in self.index_to_indices.keys()
+        }
 
-        if np.any(is_inadmissible) and optim_paras["inadmissibility_penalty"] is None:
-            warnings.warn(
-                "Some choices in the model are not admissible all the time. Thus, respy"
-                " applies a penalty to the utility for these choices which is "
-                f"{INADMISSIBILITY_PENALTY} by default. For the full solution, the "
-                "penalty only needs to be larger than all other value functions to be "
-                "effective. Choose a milder penalty for the interpolation which does "
-                "not dominate the linear interpolation model."
-            )
+    def set_convenience_objects(self):
+        """
 
-        return is_inadmissible
+        """
+        self.index_to_complex = {i: k for i, k in enumerate(self.dense_period_cores)}
+        self.index_to_core_index = {
+            i: self.dense_period_cores[self.index_to_complex[i]]
+            for i in self.index_to_complex.keys()
```

Comment:
Never use `.keys()` because you loop over the keys automatically. And see how you use the key and value in the expression whcih can be simplified with `.items()`.